gossip/hub/instance.go

```diff
@@ -0,0 +1,197 @@
+// backend Trillian log servers. The rules applied are:
+//
+// 1. The backend set must define a set of hub backends with distinct
+// (non empty) names and non empty backend specs.
+// 2. The backend specs must all be distinct.
+// 3. The hub configs must all specify a hub backend and each must be one of
+// those defined in the backend set.
+// 4. The prefixes of configured hubs must all be distinct and must not be
+// empty.
+// 5. The set of tree ids for each configured backend must be distinct.
+func ValidateHubMultiConfig(cfg *configpb.HubMultiConfig) (map[string]*configpb.HubBackend, error) {
+	// Check the backends have unique non empty names and build the map.
+	backendMap := make(map[string]*configpb.HubBackend)
+	bSpecMap := make(map[string]bool)
+	for _, backend := range cfg.Backends.Backend {
+		if len(backend.Name) == 0 {
+			return nil, fmt.Errorf("empty backend name: %v", backend)
+		}
+		if len(backend.BackendSpec) == 0 {
+			return nil, fmt.Errorf("empty backend_spec for backend: %v", backend)
+		}
+		if _, ok := backendMap[backend.Name]; ok {
+			return nil, fmt.Errorf("duplicate backend name: %v", backend)
+		}
+		if ok := bSpecMap[backend.BackendSpec]; ok {
+			return nil, fmt.Errorf("duplicate backend spec: %v", backend)
+		}
+		backendMap[backend.Name] = backend
+		bSpecMap[backend.BackendSpec] = true
+	}
+
+	// Check that hubs all reference a defined backend and there are no duplicate
+	// or empty prefixes. Apply other HubConfig specific checks.
+	hubNameMap := make(map[string]bool)
+	hubIDMap := make(map[string]bool)
+	for _, hubCfg := range cfg.HubConfigs.Config {
+		if len(hubCfg.Prefix) == 0 {
+			return nil, fmt.Errorf("hub config: empty prefix: %v", hubCfg)
+		}
+		if hubNameMap[hubCfg.Prefix] {
+			return nil, fmt.Errorf("hub config: duplicate prefix: %s: %v", hubCfg.Prefix, hubCfg)
+		}
+		if _, ok := backendMap[hubCfg.HubBackendName]; !ok {
+			return nil, fmt.Errorf("hub config: references undefined backend: %s: %v", hubCfg.HubBackendName, hubCfg)
+		}
+		hubNameMap[hubCfg.Prefix] = true
+		hubIDKey := fmt.Sprintf("%s-%d", hubCfg.HubBackendName, hubCfg.LogId)
+		if ok := hubIDMap[hubIDKey]; ok {
+			return nil, fmt.Errorf("hub config: dup tree id: %d for: %v", hubCfg.LogId, hubCfg)
+		}
+		hubIDMap[hubIDKey] = true
+	}
+
+	return backendMap, nil
+}
+
+// InstanceOptions describes the options for a hub instance.
+type InstanceOptions struct {
+	Deadline      time.Duration
+	MaxGetEntries int64
+	MetricFactory monitoring.MetricFactory
+	// ErrorMapper converts an error from an RPC request to an HTTP status, plus
+	// a boolean to indicate whether the conversion succeeded.
+	ErrorMapper func(error) (int, bool)
+}
+
+// SetUpInstance sets up a hub instance that uses the specified client to communicate
+// with the Trillian RPC back end.
+func SetUpInstance(ctx context.Context, client trillian.TrillianLogClient, cfg *configpb.HubConfig, opts InstanceOptions) (*PathHandlers, error) {
```

Comment:
It would be nice if this returned a struct that implemented an interface that defined the server. 
This would allow alternate implementations of the server, testing, or otherwise.

Reply:
I'm not a fan of defining `interface`s in anticipation of them possibly being used in future (and the Go world seems to mostly agree with me -- cf. "[golang interface pollution](https://www.google.com/search?q=golang+interface+pollution)"), and alternate implementations haven't been needed for testing as yet.